### PR TITLE
T191587 - Fix error with IP users

### PR DESCRIPTION
**Pull request by dayllanmaza**

### CODE REVIEW

File: server/src/Service/InteractionService.php [hunk 1]
```diff
@@ -43,17 +43,20 @@ public function getInteraction(
...
 
-		// get user ids from usernames
-		$userIds = array_map( function ( $username ) {
-			$userId = $this->userDao->getUserId( $username );
-			if ( !$userId ) {
-				throw new \InvalidArgumentException( sprintf( 'username %s could not be found', $username ) );
-			}
-			return $userId;
-		}, $users );
+		// If none of the users is an IP, get their user ID(s) to improve performance
```
Comment (MusikAnimal) [inline server/src/Service/InteractionService.php:46]: Just curious if you have done any testing that shows `rev_user` is faster? We recently added IP support to XTools, and for this I simply rewrote most queries to use `rev_user_text` regardless of whether the inputted user was an IP or account, and I didn't notice any difference in performance. When I did EXPLAIN on the queries (on the replicas), I did see slightly fewer estimated rows with `rev_user`, but again even for users with a lot of edits the run time seems to be roughly the same.

Comment (dayllanmaza) [reply server/src/Service/InteractionService.php:46]: I remember it was significantly faster while I was working on it. Mostly the query for common pages. I'll run some examples tomorrow and get some proper benchmarks for you

File: server/src/Dao/RevisionDao.php [hunk 2]
```diff
@@ -110,4 +112,22 @@ public function getRevisionDetails( $revIds ) {
...
+	 *
+	 * When possible, search for user ids to improve performance.
+	 * If usernames or IP(s) are provided, then use those.
+	 *
+	 * @param array $users List of usernames/IP(s), ids, or mixed
+	 * @return string
+	 */
+	private function getUserSearchColumn( $users ) {
+		$column = 'rev_user_text';
+		if ( array_filter( $users, 'is_numeric' ) ) {
```
Comment (davidbarratt) [inline server/src/Dao/RevisionDao.php]: I'm assuming IPv4 and IPv6 do not pass `is_numeric` ? Also, this doesn't seem to enforce that _all_ of them are numeric, just that at least one of them are.

File: server/src/Dao/RevisionDao.php [hunk 3]
```diff
@@ -110,4 +112,22 @@ public function getRevisionDetails( $revIds ) {
...
 		return $this->fetchAll( $query, \PDO::FETCH_ASSOC );
 	}
+
+	/**
+	 * Determine user's search column on revisions table.
+	 *
+	 * When possible, search for user ids to improve performance.
+	 * If usernames or IP(s) are provided, then use those.
+	 *
+	 * @param array $users List of usernames/IP(s), ids, or mixed
```
Comment (davidbarratt) [inline server/src/Dao/RevisionDao.php]: It doesn't seem liked `mixed` is actually allowed, as far as I can tell, it has to be a list of one or the other (a list of "usernames/IP(s) or user ids"? or does `rev_user_text` also contain the user id?

**Review by davidbarratt — CHANGES_REQUESTED** [review]

File: server/src/Dao/RevisionDao.php (hunk 3 above)
Comment (dayllanmaza) [reply server/src/Dao/RevisionDao.php]: No it does not allow mixed. Let me fix that

File: server/src/Dao/RevisionDao.php (hunk 2 above)
Comment (dayllanmaza) [reply server/src/Dao/RevisionDao.php]: Woopsie.. I guess it works 'cause by the time it gets here it is either all ids or usernames. Let me make this more restrictive.
And IPv4/IPv5 do not pass `is_numeric`

File: server/src/Dao/RevisionDao.php [hunk 4]
```diff
@@ -110,4 +112,38 @@ public function getRevisionDetails( $revIds ) {
...
+		return $column;
+	}
+
+	/**
+	 * Determine if all the elements in an array are numeric.
+	 *
+	 * @param array $arr
+	 * @return boolean
+	 */
+	private function isNumericArray( $arr ) {
```
Comment (davidbarratt) [inline server/src/Dao/RevisionDao.php]: you have a _similar_ implementation in `isIpPresent` but uses a completely different methodology (loop vs filter). I prefer the filter, but either way, I think it should be consistent (especially in the same PR, ha).

Also, add an `array` typehint.

File: server/src/Service/InteractionService.php [hunk 5]
```diff
@@ -142,4 +145,18 @@ private function sanitizeDate( $date ) {
...
+	 *
+	 * @param array $data
+	 * @return boolean
+	 */
+	private function isIpPresent( $data ) {
+		$ips = array_filter( $data, function ( $possibleIp ) {
+			return filter_var( $possibleIp, FILTER_VALIDATE_IP );
+		} );
+
+		return count( $ips ) > 0;
```
Comment (davidbarratt) [inline server/src/Service/InteractionService.php]: nit-pick: I think you could just do something like:
```php
return (boolean) array_filter( $data, function ( $possibleIp ) {
  return filter_var( $possibleIp, FILTER_VALIDATE_IP );
} );
```
but if that's harder to read, then what you have is fine.

File: server/src/Service/InteractionService.php [hunk 6]
```diff
@@ -142,4 +145,18 @@ private function sanitizeDate( $date ) {
...
 		return isset( $dateObj ) ? $dateObj : null;
 	}
+
+	/**
+	 * Check if a value in an array is a valid IP address
+	 *
+	 * @param array $data
+	 * @return boolean
+	 */
+	private function isIpPresent( $data ) {
```
Comment (davidbarratt) [inline server/src/Service/InteractionService.php]: add an `array` typehint
https://secure.php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration

File: server/src/Dao/RevisionDao.php [hunk 7]
```diff
@@ -110,4 +112,38 @@ public function getRevisionDetails( $revIds ) {
...
+	/**
+	 * Determine user's search column on revisions table.
+	 *
+	 * When possible, search for user ids to improve performance.
+	 * If usernames or IP(s) are provided, then use those.
+	 *
+	 * @param array $users List of usernames/IPs or user ids
+	 * @return string
+	 */
+	private function getUserSearchColumn( $users ) {
```
Comment (davidbarratt) [inline server/src/Dao/RevisionDao.php]: Add an `array` typehint
https://secure.php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration

File: server/src/Dao/RevisionDao.php [hunk 8]
```diff
@@ -9,7 +9,7 @@ class RevisionDao extends AbstractDao {
 	/**
 	 * Fetch user interaction in commonly edited pages
 	 *
-	 * @param int[] $users
+	 * @param array $users List of usernames/IP(s), ids, or mixed
```
Comment (davidbarratt) [inline server/src/Dao/RevisionDao.php]: I don't think this supports `mixed`?
See the docblock on `getUserSearchColum()`

File: server/src/Dao/RevisionDao.php [hunk 9]
```diff
@@ -50,7 +51,7 @@ public function getUserRevisionsInCommonPages(
 	/**
 	 * Get common edited pages between multiple users
 	 *
-	 * @param int[] $users
+	 * @param array $users List of usernames/IP(s), ids, or mixed
```
Comment (davidbarratt) [inline server/src/Dao/RevisionDao.php]: I don't think this supports `mixed`?
See the docblock on `getUserSearchColum()`

**Review by davidbarratt — CHANGES_REQUESTED** [review]

File: server/src/Dao/RevisionDao.php (hunk 4 above)
Comment (dayllanmaza) [reply server/src/Dao/RevisionDao.php]: It doesn't make sense to use array_filter here. I don't need to loop through all the elements.

File: server/src/Service/InteractionService.php (hunk 5 above)
Comment (dayllanmaza) [reply server/src/Service/InteractionService.php]: I think count is clearer tho. `array_filter` returns an array and casting an array to boolean feels weird to me.

File: server/src/Service/InteractionService.php (hunk 6 above)
Comment (dayllanmaza) [reply server/src/Service/InteractionService.php]: :+1:

File: server/src/Dao/RevisionDao.php (hunk 7 above)
Comment (dayllanmaza) [reply server/src/Dao/RevisionDao.php]: :+1:

File: server/src/Dao/RevisionDao.php (hunk 4 above)
Comment (davidbarratt) [reply server/src/Dao/RevisionDao.php]: I don't think you need to loop through all of the elements in `isIpPresent` either? Shouldn't it just return as soon as the first IP is found?

Comment (dayllanmaza) [reply server/src/Dao/RevisionDao.php]: fixed

**Review by davidbarratt — APPROVED** [review]

**Pull request merged**